File: Linked_List/025.Reverse-Nodes-in-k-Group/025.Reverse-Nodes-in-k-Group.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseKGroup(ListNode* head, int k) 
    {
        if (k==1) return head;
        
        ListNode* dummy = new ListNode();
        dummy->next = head;
        
        ListNode* p = dummy;
        while (p)
        {
            ListNode* q = p;
            int count = 0;
            while (q->next && count < k)
            {
                q = q->next;
                count++;
            }
            if (count < k)
                return dummy->next;
            
            ListNode* cur = p->next;
            ListNode* nxt = cur->next;
            ListNode* future = nxt->next;
            ListNode* start = cur;
            
            for (int i=0; i<k-1; i++)
            {
                nxt->next = cur;
                cur = nxt;
                nxt = future;
                if (future) future = future->next;
            }
            p->next = cur;
            start->next = nxt;
            
            p = start;
        }
        
        return dummy->next;        
    }
};

File: Linked_List/025.Reverse-Nodes-in-k-Group/025.Reverse-Nodes-in-k-Group.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def reverseKGroup(self, head, k):
        """
        :type head: ListNode
        :type k: int
        :rtype: ListNode
        """
        dummyHead = ListNode(0)
        dummyHead.next = head
        h = dummyHead
        while h.next:
            count = 0
            left = h
            while (h.next and count<k):
                h = h.next
                count+=1
            if count<k: return dummyHead.next
            right = h.next
            
            a,b = left.next, left.next.next
            for i in range(k-1):                
                temp = b.next
                b.next = a
                a,b = b, temp                
                            
            h = left.next
            left.next = a
            h.next = right
        return dummyHead.next
            
